# Prevents using app.listen with 0.0.0.0 locally (app-listen-locally)

Rather than using `app.listen(PORT, () => {})` which automatically binds to `0.0.0.0`, let us bind to `127.0.0.1` when testing locally.

## Rule Details

This rule aims to prevent uses of `app.listen(PORT, () => {})` with default `0.0.0.0` interface. 

This rule can automatically fix such usage to be use `process.env._IS_LOCAL` as a way to determine local run vs not.
```js
app.listen(PORT, () => {});
```
The above code will be fixed into.
```js
const isLocal = process.env._IS_LOCAL === "true";
app.listen(PORT, isLocal ? "localhost" : "0.0.0.0", () => {});
```


Examples of **incorrect** code for this rule:


```js
app.listen(port, '127.0.0.1', () => console.log('Testing' + port));
app.listen(port, '0.0.0.0', () => console.log('Testing' + port));
app.listen(port, () => console.log('Testing' + port));
```

Examples of **correct** code for this rule (repo-specific):

```js
app.listen(port, isLocal ? "localhost" : "0.0.0.0", () => console.log('Testing' + port));

const myhostvar = isLocal ? "localhost" : "0.0.0.0"; 
app.listen(port, myhostvar, () => console.log('Testing' + port));
```

## When Not To Use It

There may be some cases where this rule should be disabled. 
