# Prevents supplying functions in mapStateToProps return object (no-functions-in-map-state-to-props)

Supplying functions in the object returned by `mapStateToProps` can have the undesirable side effect of causing extra re-renders by React. In a nutshell:
1. Redux store update occurs
2. `mapStateToProps` is called on all store-connected components
3. `mapStateToProps` returns a function as a prop
4. React compares old function prop to new function prop, and finds a difference
5. React re-renders

This results in the behavior where for every store update, the offending components will always re-render.

## Rule Details

This rule seeks to identify functions that are returned in the `mapStateToProps` return object.

Note: Currently, this lint rule only finds anonymous arrow functions, but in the future, this rule should be extended to catch any function.

Examples of **incorrect** code for this rule:

```js
function mapStateToProps(state) {
  return {
    key: () => {},
  };
}
```
```js
const mapStateToProps = (state) => ({
  finder: (id) => selectors.objByID(state, id),
});
```

Examples of **correct** code for this rule (repo-specific):

```js
function mapStateToProps(state) {
  return {
    key: selectors.someSelector(state),
    finder: selectors.allObjByID(state),
  };
}
```

## When Not To Use It

N/A. Returning functions in `mapStateToProps` should be avoided in all cases.
