# Prevents using res.status and res.sendStatus with an error status code (no-send-status-error)

Rather than using `res.status` or `res.sendStatus` for status codes in the 4XX-5XX range, a common practice at Clever is to throw an error which will get handled by the Express error handler we set up. These errors vary by repo, but often look like `NotFoundError` or `ErrorWithStatusCode`. The benefit of this approach is that it creates a central place for error handling where we can do any desired logging or reporting. It can also make it easier to bubble up errors from libs used by the server.

## Rule Details

This rule aims to prevent uses of `res.status` and `res.sendStatus` with error status codes, instead prompting the engineer to throw the appropriate error instead.

Examples of **incorrect** code for this rule:

```js
res.sendStatus(404);
res.status(401).json({ message: "User not logged in", code: ErrorCode.NO_PORTAL_SESSION });
```

Examples of **correct** code for this rule (repo-specific):

```js
throw new NotFoundError();
throw new AuthorizationError("User not logged in", ErrorCode.NO_PORTAL_SESSION);
```

## When Not To Use It

There may be some cases where this rule should be disabled. In fact, every repo will have at least one: the error handler. This is because the error handler is the central place where we convert the thrown errors into status codes. Don't hesitate to disable this rule for the error handler or other appropriate cases.
